libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/KeyPair.hs

```diff
@@ -110,15 +115,53 @@ deriving instance Typeable r => Eq (KeyPair r)
 instance ToExpr (KeyPair r) where
   toExpr (KeyPair x y) = Tree.App "KeyPair" [toExpr x, Tree.App (take 10 (show y)) []]
 
-mkAddr :: (KeyPair 'Payment, KeyPair 'Staking) -> Addr
-mkAddr (payKey, stakeKey) = Addr Testnet (mkCred payKey) (StakeRefBase $ mkCred stakeKey)
+class MakeCredential c r where
+  mkCred :: c -> Credential r
+instance MakeCredential (Credential r) r where
+  mkCred = id
+instance MakeCredential (KeyPair r) r where
+  mkCred = KeyHashObj . hashKey . vKey
+instance MakeCredential (KeyHash r) r where
+  mkCred = KeyHashObj
+instance MakeCredential ScriptHash r where
+  mkCred = ScriptHashObj
+
+class MakeStakeReference c where
+  mkStakeRef :: c -> StakeReference
+  default mkStakeRef :: MakeCredential c Staking => c -> StakeReference
+  mkStakeRef = StakeRefBase . mkCred
+instance MakeStakeReference StakeReference where
+  mkStakeRef = id
+instance MakeStakeReference (Credential Staking)
+instance MakeStakeReference (KeyPair Staking)
+instance MakeStakeReference (KeyHash Staking)
+instance MakeStakeReference ScriptHash
+instance MakeStakeReference Ptr where
+  mkStakeRef = StakeRefPtr
+instance MakeStakeReference (Maybe StakeReference) where
+  mkStakeRef = mkStakeRefMaybe
+instance MakeStakeReference (Maybe (Credential Staking)) where
+  mkStakeRef = mkStakeRefMaybe
+instance MakeStakeReference (Maybe (KeyPair Staking)) where
+  mkStakeRef = mkStakeRefMaybe
+instance MakeStakeReference (Maybe (KeyHash Staking)) where
+  mkStakeRef = mkStakeRefMaybe
+instance MakeStakeReference (Maybe ScriptHash) where
+  mkStakeRef = mkStakeRefMaybe
+
+mkStakeRefMaybe :: MakeStakeReference c => Maybe c -> StakeReference
+mkStakeRefMaybe = \case
+  Nothing -> StakeRefNull
+  Just c -> mkStakeRef c
+
+-- | Construct a `Testnet` address from payment and staking components
+mkAddr :: (MakeCredential p 'Payment, MakeStakeReference s) => p -> s -> Addr
+mkAddr pay stake = Addr Testnet (mkCred pay) (mkStakeRef stake)
 
 mkScriptAddr :: ScriptHash -> KeyPair 'Staking -> Addr
 mkScriptAddr scriptHash stakeKey =
   Addr Testnet (ScriptHashObj scriptHash) (StakeRefBase $ mkCred stakeKey)
-
-mkCred :: KeyPair kr -> Credential kr
-mkCred k = KeyHashObj . hashKey $ vKey k
+{-# DEPRECATED mkScriptAddr "In favor of `mkAddr`" #-}
```

Comment:
I think you meant to refer to `MakeCredential` here
```suggestion
{-# DEPRECATED mkScriptAddr "In favor of `mkAddr` method in `MakeCredential`" #-}
```

Reply:
`mkAddr` is not in `MakeCredential` type class. It is a standalone function

Reply:
Oh, ok, I thought it applies to `mkCred`, I misread the diff